src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -462,6 +514,34 @@ private void AddIfCaptured(Symbol symbol)
                 }
 
                 /// <summary>
+                /// Add a diagnostic if the type of a captured variable is a restricted type
+                /// </summary>
+                private void AddDiagnosticIfRestrictedType(Symbol capturedVariable, SyntaxNode syntax)
+                {
+                    TypeSymbol type;
+                    switch (capturedVariable.Kind)
+                    {
+                        case SymbolKind.Local:
+                            type = ((LocalSymbol)capturedVariable).Type;
+                            break;
+                        case SymbolKind.Parameter:
+                            type = ((ParameterSymbol)capturedVariable).Type;
+                            break;
+                        default:
+                            // This should only be called for captured variables, and captured
+                            // variables must be a method, parameter, or local symbol
+                            Debug.Assert(capturedVariable.Kind == SymbolKind.Method);
+                            type = null;
+                            break;
+                    }
+
+                    if (type?.IsRestrictedType() == true)
+                    {
+                        _diagnostics.Add(ErrorCode.ERR_SpecialByRefInLambda, syntax.Location, type);
```

Comment:
Off-topic: Has this error message been updated to include "local functions"? I see the deleted equivalent for this in the old code has a comment saying it lists "anonymous function, query expression, iterator block or async method", but that comment may be out of date (and it's been fixed already).